File: root_leaf_sum.py
#time - n,space - h
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.sum=0
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        self.helper(root,root.val)
        return self.sum
    def helper(self,root,s):
        #base
        if root.left==None and root.right==None:
            self.sum+=s
            return
        #logic
        if root.left:
            self.helper(root.left,(s*10)+root.left.val)
        if root.right:
            self.helper(root.right,(s*10)+root.right.val)

File: RootLeafSum.java
// Time Complexity : O(N)
// Space Complexity : O(H)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach

class Solution {

    int total = 0;

    private void helper(TreeNode root, int sum) {

        if (root == null) return;

        sum = sum * 10 + root.val;

        // check if its a leaf node
        if (root.left == null && root.right == null) {
            total += sum;
            return;
        }

        // Go to left and right with the sum
        helper(root.left, sum);
        helper(root.right, sum);

    }

    public int sumNumbers(TreeNode root) {

        helper(root, 0);

        return total;

    }
}